polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -431,6 +492,64 @@ pub mod pallet {
 			runtime_parachains::set_current_head::<T>(para, new_head);
 			Ok(())
 		}
+
+		/// Changes the billing account of a parachain to the caller.
+		///
+		/// ## Deposits/Fees
+		/// For this call to be successful, the caller account must have a sufficient balance
+		/// to cover the current deposit required for this parachain.
+		///
+		/// Can be called by Root, the parachain, or the parachain manager if the parachain is
+		/// unlocked.
+		#[pallet::call_index(9)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_parachain_billing_account_to_self())]
+		pub fn set_parachain_billing_account_to_self(
+			origin: OriginFor<T>,
+			para: ParaId,
+		) -> DispatchResult {
+			Self::ensure_root_para_or_owner(origin.clone(), para)?;
+
+			let new_billing_account = if let Ok(caller) = ensure_signed(origin.clone()) {
+				let para_info = Paras::<T>::get(para).ok_or(Error::<T>::NotRegistered)?;
+				ensure!(!para_info.is_locked(), Error::<T>::ParaLocked);
+				ensure!(para_info.manager == caller, Error::<T>::NotOwner);
```

Comment:
I understand your point; however, I am still unsure whether this is necessary, mainly because the `T::DataDepositPerByte` probably isn't something that is going to change frequently. Calculating the associated cost of an upgrade depends on only one parameter: the byte size of the new validation code, and two constant values (`T::DataDepositPerByte` and `T::UpgradeFee`) so it should be pretty hard to miscalculate.

However, a much simpler solution to this problem would be to add a new optional parameter to the `schedule_code_upgrade` extrinsic that specifies the maximum cost of an upgrade that the parachain is willing to pay. Before performing the upgrade, the extrinsic will just check upfront whether the required cost is under the specified maximum.